geode-for-redis/src/main/java/org/apache/geode/redis/internal/data/RedisList.java

```diff
@@ -356,10 +432,32 @@ protected synchronized void elementsPushHead(List<byte[]> elementsToAdd) {
     }
   }
 
+  public synchronized void elementsRemove(List<Integer> indexList) {
+    for (Integer element : indexList) {
+      elementList.remove(element.intValue());
+    }
+  }
+
   public synchronized void elementReplace(int index, byte[] newValue) {
     elementList.set(index, newValue);
   }
 
+  public synchronized void elementsRetainByIndexRange(int start, int end) {
+    if (start < 0) {
+      // Remove everything
+      elementList.clear();
+      return;
+    }
+
+    if (end < elementList.size()) {
+      elementList.subList(end + 1, elementList.size()).clear();
+    }
+
+    if (start > 0) {
+      elementList.subList(0, start).clear();
+    }
```

Comment:
When a delta comes in, we may remove things at the beginning of the list, or the end, or both. They are separate calls to subList().clear(), and I'm trying to avoid calling them if there's no need. If the delta values are invalid, versioning is _supposed_ to catch that...